### PR TITLE
Tp53/NF1 classifier: Evaluate classifier using CNV alterations

### PR DESCRIPTION
**Pull request by kgaonkar6**
### Purpose/implementation Section 
Update TP53/NF1 alterations that are used in the classifier evaluation step for analysis: [tp53_nf1_score](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/tp53_nf1_score)

#### What scientific question is your analysis addressing?
We used SNVs to get tp53/nf1 altered status for evaluating the classifier results, should we add CNV calls as well to identify tp53/nf1 deletion inactivation status of samples as per #165 (comment) ?

#### What was your approach?
Add TP53/NF1 loss status from data/consensus_seg_annotated_cn_autosomes.tsv.gz to results/TP53_NF1_snv_alteration.tsv 

#### What GitHub issue does your pull request address?
#653 

#### Which areas should receive a particularly close look?
TP53/NF1 loss status selection from data/consensus_seg_annotated_cn_autosomes.tsv.gz
[here](https://github.com/kgaonkar6/OpenPBTA-analysis/blob/87360e83ec6d4030e6fa9015d32a4fa1589a48d1/analyses/tp53_nf1_score/00-tp53-nf1-alterations.R#L83)
And "No_Tp53_NF1_alt" status assignment [here](https://github.com/kgaonkar6/OpenPBTA-analysis/blob/87360e83ec6d4030e6fa9015d32a4fa1589a48d1/analyses/tp53_nf1_score/00-tp53-nf1-alterations.R#L125)

#### Is there anything that you want to discuss further?
- I've assigned "No_TP53_NF1_alt" to TP53 amplification/gain status since we are looking for TP53/NF1 loss conditions. 
- Bs_ids which have neutral TP53/NF1 status are not in the consensus_seg_annotated_cn_autosomes.tsv.gz these are also annotated as "No_TP53_NF1_alt.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes

#### What types of results are included (e.g., table, figure)?
figures

#### What is your summary of the results?
Results might be updated, it seems the addition of all "losses" is not a good match to evaluate looking at the ROC curve for stranded data here: 
stranded_TP53
stranded_NF1

#### Reproducibility Checklist

- [X] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [X] This analysis has been added to continuous integration.

#### Documentation Checklist

- [ ] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by jaclyn-taroni** [comment]
Hi @kgaonkar6 - over on #653, we talked about the "permissiveness" of the annotations (thinking of https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/653#issuecomment-610518874, specifically). Some changes got merged into [`focal-cn-file-preparation`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/focal-cn-file-preparation) such that there is now a file that accounts for broad CNA and is a table of the "most focal" unit (arm CNA vs. cytoband CNA vs. gene CNA). That file is: `analyses/focal-cn-file-preparation/results/consensus_seg_most_focal_cn_status.tsv.gz`. Naively, I expect that the broader calls might be a little more robust and we might want to try looking at the cytoband level for _TP53_ losses.

More generally, though - what are we hoping to gain or understand from the classifier analysis? (Note: I filed #165 because we were contacted via email about doing that analysis.) This could end up being a bit of a rabbit hole and I am trying to weigh this against tasks like #665.

**Comment by kgaonkar6** [comment]
I was just trying to close loose ends with this PR :D . I'm not sure how it weighs in preferences to the BRAF fusion analysis . 

The  aim of running the classifier was to identify if there are samples which that have high tp53 inactivation score ( thus predicted to have inactivated tp53 profile) where maybe TP53 is not actually mutated or deleted or we don't have the dna data to look at that.

**Comment by jaclyn-taroni** [comment]
> I was just trying to close loose ends with this PR :D .

Yea, I gotcha. Worth taking a step back, though!

> The aim of running the classifier was to identify if there are samples which that have high tp53 inactivation score ( thus predicted to have inactivated tp53 profile) where maybe TP53 is not actually mutated or deleted or we don't have the dna data to look at that.

So I think the handling of ploidy across all the copy number calls is somewhat inconsistent, which concerns me when it comes to these updates. 

Would it be more interesting, instead, to start by looking at negative (e.g., non-mutated) cases of _TP53_ that have the highest inactivation scores?

**Comment by kgaonkar6** [comment]
> Worth taking a step back, though!

Right let me know if you think this can be de-prioritized as well .

> Would it be more interesting, instead, to start by looking at negative (e.g., non-mutated) cases of _TP53_ that have the highest inactivation scores?

yeah I think that would be interesting !

**Comment by cgreene** [comment]
"Would it be more interesting, instead, to start by looking at negative (e.g., non-mutated) cases of TP53 that have the highest inactivation scores?"

We have had really good luck with these types of approaches revealing interesting biology in the past! Strong agree on exploring the top set of these. What are the confidences like? Moving classifiers out of the setting in which they are trained can often lead to probabilities that are a bit compressed towards the middle.

**Comment by kgaonkar6** [comment]
Oh yeah I did work on the confidence intervals for ROC, I'll add that code to this PR as well.
ROC for the tp53_scores
ci_roc_tp53

ROC for nf1_scores
ci_roc_NF1

**Comment by jaclyn-taroni** [comment]
@cgreene when you say 

> What are the confidences like? Moving classifiers out of the setting in which they are trained can often lead to probabilities that are a bit compressed towards the middle.

Are you looking for the confidence intervals for ROC or are you referring to something else?

**Comment by cgreene** [comment]
The distribution of probabilities from the logistic regression classifier. I'm guessing that instead of running from 0-1 they'll be more narrowly distributed around the middle.

**Comment by kgaonkar6** [comment]
I'm no expert in classifiers but from the code it looks like there is only a logit transformed score https://github.com/kgaonkar6/OpenPBTA-analysis/blob/a203b23478b9b3bb0ed1f19b78b2b04342f0bf6c/analyses/tp53_nf1_score/utils.py#L47 

Looks like p value can be calculated as such : 1/(1+exp(L)) ?

https://data-se.netlify.app/2017/01/24/convert_logit2prob/
Screen Shot 2020-05-22 at 9 58 27 AM

**Comment by cgreene** [comment]
@kgaonkar6 isn't that what this line already does? https://github.com/kgaonkar6/OpenPBTA-analysis/blob/a203b23478b9b3bb0ed1f19b78b2b04342f0bf6c/analyses/tp53_nf1_score/utils.py#L49

I think you can just plot the output of the classifier.

**Comment by kgaonkar6** [comment]
oh yeah that is what the scores are, ok I'll plot those!

**Comment by kgaonkar6** [comment]
Boxplots for scores does look like mostly distributed around the 0.5 mark:

polya_score_dist
stranded_score_dist

**Comment by cgreene** [comment]
@kgaonkar6 : can you plot separately by genotyping-derived status too? TP53 only, unless you want to do NF1. It doesn't look like NF1 is working in this setting from the results above.

**Comment by kgaonkar6** [comment]
The tp53_status as x axis in the plots below are derived from the genomic data (1= has tp53 alteration and 0= doesn't have tp53 alterations)

polya_score_dist
TP53 status | freq
--- | ---
1   | 29
0   | 29

stranded_score_dist
TP53 status |freq
--- | ---
0   | 685
1   | 285

**Comment by cgreene** [comment]
Ok! Yes - this makes a lot of sense. There is a difference by TP53 status and it goes in the anticipated direction - so the classifier seems to be porting successfully to this setting. On the other hand, the _probabilities_ coming out of the classifier look like they're likely to be a bit compressed.

You could make a calibration curve ( description if helpful: https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_curve.html ) if we really want to dig into understanding if the classifier is under- or over-confident in this setting. Based on what I see here, my guess is that it's under-confident in the poly-A setting and maybe overconfident in the stranded setting.

I don't think that's really necessary though, if we consider the results that come out as a ranked list instead of probabilities. If you took the top "false positives" (those with 0 status but with very high confidence scores), it could help to investigate those to see if they fall into certain tumor types, have certain mutations, or have other evidence of TP53 loss. In other settings we've found this type of analysis quite fruitful.

**Comment by kgaonkar6** [comment]
> If you took the top "false positives" (those with 0 status but with very high confidence scores), it could help to investigate those to see if they fall into certain tumor types, have certain mutations, or have other evidence of TP53 loss. In other settings we've found this type of analysis quite fruitful.

Sounds good! I can start by looking into top 100 "false positives" from stranded dataset.

**Comment by jaclyn-taroni** [comment]
@kgaonkar6 do you want to close this pull request and work on the ranked list stuff in a new one? Not sure if those boxplots included CNV alterations. (I'm after how relevant what your next steps will be are to #653.)

**Comment by kgaonkar6** [comment]
I did include the CNV alterations (TP53 "loss" from consensus_seg_annotated_cn_autosomes.tsv.gz ) tp53_status counts in the boxplots above, would you suggest only using SNVs for now ?

If we are only interested in using SNV alterations then I'll close this PR like you said, since the next steps ie. looking at top false positives should probably be a different PR.

**Comment by jaclyn-taroni** [comment]
> I did include the CNV alterations (TP53 "loss" from consensus_seg_annotated_cn_autosomes.tsv.gz ) tp53_status counts in the boxplots above, would you suggest only using SNVs for now ?

Nope, not necessarily -  I would be interested in the difference in scores between _TP53_ losses and _TP53_ coding mutations. In that case, I will review this - including the poor performance for ROC seems appropriate. I'd like to see those boxplots make it into the repo, too, but that can be a separate PR.

**Comment by kgaonkar6** [comment]
Thanks Jaclyn! I'll work on the next PR .
